apps/labs/posts/narwhals-pycapsule.md

```diff
@@ -0,0 +1,199 @@
+- [Narwhals](https://github.com/narwhals-dev/narwhals), if you want to keep your logic in Python.
+- Narwhals _and_ the PyCapsule Interface together, if you want it all!
+
+## Summing non-null values in a column - "slow down, Professor!"
+
+We'll learn about how you make a dataframe-agnostic function `agnostic_sum_i64_column`.
+Here are the requirements:
+
+- Given a dataframe `df` and a column name `column_name`, compute
+  the sum of all non-null values in that column.
+- If that column is not present, or if it is not of type `Int64`,
+  raise an error.
+- `pandas.DataFrame`, `polars.DataFrame`,
+  `duckdb.PyRelation`, and `pyarrow.Table`, should all be supported without
+  any of them being required.
+
+Summing non-null values isn't rocket science, but how can we do this in a way that all
+input libraries are supported without any of them being required?
+
+## Low-level solution: PyCapsule Interface in Rust via PyO3
+
+An example of a Rust solution via Pyo3 can be found at [pycapsule-demo/src/lib.rs](https://github.com/MarcoGorelli/pycapsule-demo/blob/6aad64be26ebbfc8526f26695544bfc6436e3266/src/lib.rs#L9-L56).
+
+The technical details are beyond the scope of this post, but the summary is:
+
+- We accept any object which implements the `ArrowStreamExportable` protocol from the
+  [PyCapsule Interface](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html).
+  We can check for this by looking for the `__arrow_c_stream__` attribute:
+
+  ```python
+  >>> import polars as pl
+  >>> import duckdb
+  >>> import pandas as pd
+  >>> import pyarrow as pa
+  >>> hasattr(pl.DataFrame, '__arrow_c_stream__')
+  True
+  >>> hasattr(duckdb.DuckDBPyRelation, '__arrow_c_stream__')
+  True
+  >>> hasattr(pd.DataFrame, '__arrow_c_stream__')
+  True
+  >>> hasattr(pa.Table, '__arrow_c_stream__')
+  True
+  ```
+- The solution is totally agnostic to the exact input, as it respects a standardised interface.
+  We can verify this by trying to pass different dataframe libraries to the function:
+
+  ```python
+  >>> from pycapsule_demo import agnostic_sum_i64_column
+  >>> df_pl = pl.DataFrame({'a': [1,1,2], 'b': [4,5,6]})
+  >>> df_pd = pd.DataFrame({'a': [1,1,2], 'b': [4,5,6]})
+  >>> df_pa = pa.table({'a': [1,1,2], 'b': [4,5,6]})
+  >>> rel = duckdb.sql("select * from df_pd")
+  >>> agnostic_sum_i64_column(df_pl, column_name="a")
+  6
+  >>> agnostic_sum_i64_column(df_pd, column_name="a")
+  6
+  >>> agnostic_sum_i64_column(df_pa, column_name="a")
+  6
+  >>> agnostic_sum_i64_column(rel, column_name="a")
+  6
+  ```
+
+Like magic, our function works agnostically for any of these dataframes,
+without us having to write any specialised code to handle the subtle differences
+between them!
+
+> **_NOTE:_** How many times can `agnostic_sum_i64_column` be called? Strictly
+> speaking, the PyCapsule Interface only guarantees that it can be called once for
+> any given input. In practice, however, all implementations that we're aware of
```

Comment:
FWIW arro3 IO functions like [`read_parquet`](https://kylebarron.dev/arro3/latest/api/io/#arro3.io.read_parquet) I think will provision a single stream instance. So the first time you use the stream pointer from `read_parquet` it would consume the stream I think and you wouldn't be able to use it again.